ddl/ddl_api.go

```diff
@@ -1096,6 +1096,68 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func setTableAutoRandomBits(tbInfo *model.TableInfo, colDefs []*ast.ColumnDef) error {
+	allowAutoRandom := config.GetGlobalConfig().Experimental.AllowAutoRandom
+	pkColName := tbInfo.GetPkName()
+	return forEachColWithAutoRandomOpt(colDefs, func(col *ast.ColumnDef, autoRandomBits uint64) error {
+		if !allowAutoRandom {
+			return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomExperimentalDisabledErrMsg)
+		}
+
+		if tbInfo.PKIsHandle && col.Name.Name.L == pkColName.L {
+			if containsColumnOption(col, ast.ColumnOptionAutoIncrement) {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomIncompatibleWithAutoIncErrMsg)
+			}
+			if containsColumnOption(col, ast.ColumnOptionDefaultValue) {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomIncompatibleWithDefaultValueErrMsg)
+			}
+			fieldTypeBitsLength := uint64(mysql.DefaultLengthOfMysqlTypes[col.Tp.Tp] * 8)
+			if autoRandomBits == 0 {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(autoid.AutoRandomNonPositive)
+			} else if autoRandomBits >= fieldTypeBitsLength {
+				return ErrInvalidAutoRandom.GenWithStackByArgs(fmt.Sprintf(autoid.AutoRandomOverflowErrMsg, autoRandomBits, fieldTypeBitsLength))
+			}
+			tbInfo.AutoRandomBits = autoRandomBits
+			return nil
+		}
+		return ErrInvalidAutoRandom.GenWithStackByArgs(fmt.Sprintf(autoid.AutoRandomPKisNotHandleErrMsg, col.Name.Name.O))
+	})
+}
+
+func forEachColWithAutoRandomOpt(colDefs []*ast.ColumnDef, fn func(colDef *ast.ColumnDef, autoRandomBits uint64) error) error {
+	for _, col := range colDefs {
+		if containsColumnOption(col, ast.ColumnOptionAutoRandom) {
+			autoRandBits := getAutoRandomBits(col)
+			uAutoRandBits, err := convertAutoRandomBitsToUnsigned(autoRandBits)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err := fn(col, uAutoRandBits); err != nil {
```

Comment:
Could we use a general function here？ I think it will be more readable.

Reply:
Changing to a general function will lead to a 5-layer nested structure. Is it acceptable?

Reply:
I don't understand. We only need to replace `fn` with `XXXFunc`.